migrations/20220607041349_add_mfa_schema.up.sql

```diff
@@ -0,0 +1,51 @@
+-- See: https://stackoverflow.com/questions/7624919/check-if-a-user-defined-type-already-exists-in-postgresql/48382296#48382296
+DO $$ BEGIN
+    CREATE TYPE factor_type AS ENUM('totp', 'webauthn');
+    CREATE TYPE factor_status AS ENUM('disabled', 'unverified', 'verified');
+EXCEPTION
+    WHEN duplicate_object THEN null;
+END $$;
+
+-- auth.mfa_factors definition
+CREATE TABLE IF NOT EXISTS auth.mfa_factors(
+       id VARCHAR(256) NOT NULL,
+       user_id uuid NOT NULL,
+       friendly_name VARCHAR(256) NULL,
+       factor_type factor_type NOT NULL,
+       status factor_status NOT NULL,
+       created_at timestamptz NOT NULL,
+       updated_at timestamptz NOT NULL,
+       secret_key VARCHAR(256) NOT NULL,
+       UNIQUE(user_id, friendly_name),
+       CONSTRAINT mfa_factors_pkey PRIMARY KEY(id),
+       CONSTRAINT mfa_factors_user_id_fkey FOREIGN KEY (user_id) REFERENCES auth.users(id) ON DELETE CASCADE
+);
+comment on table auth.mfa_factors is 'Auth: stores metadata about factors';
+
+-- auth.mfa_challenges definition
+CREATE TABLE IF NOT EXISTS auth.mfa_challenges(
+       id VARCHAR(256) NOT NULL,
+       factor_id VARCHAR(256) NOT NULL,
+       created_at timestamptz NOT NULL,
+       verified BOOLEAN NOT NULL,
+       CONSTRAINT mfa_challenges_pkey PRIMARY KEY (id),
+       CONSTRAINT mfa_challenges_auth_factor_id_fkey FOREIGN KEY (factor_id) REFERENCES auth.mfa_factors(id) ON DELETE CASCADE
+);
+comment on table auth.mfa_challenges is 'Auth: stores metadata about challenge requests made';
+
+-- auth.mfa_recovery_codes definition
+CREATE TABLE IF NOT EXISTS auth.mfa_recovery_codes(
+       id serial PRIMARY KEY,
```

Comment:
just wondering, why is this serial while the other tables are using varchar(256) for the id type?

Reply:
also, if you wanna go with serial, you should consider using `bigint generated always as identity` since this would accommodate up to 8bytes (similar to `varchar(256)` but autoincrementing)

Reply:
Yup great question -- the other two are meant to be of the form `prefix_<uuid>` (e.g. `challenge_1231524`) since they are often used together in client lib/endpoint calls and it's purely to help the dev distinguish between the various uuids.

For recovery codes the id is not likely to be used elsewhere/exposed to user so there haven't been any measures. Open to changing it to varchar too though